test/shared/src/main/scala/zio/test/mock/Expectation.scala

```diff
@@ -16,9 +16,215 @@
+            out <- popNextExpectation.flatMap {
+                    case Some(next) => extract(state, next(()))
+                    case None       => UIO.succeed(Right(()))
+                  }
+          } yield out
+      }
+    }
+
+    val makeState =
+      for {
+        callsRef <- Ref.make(List.empty[AnyCall])
+        nextRef  <- Ref.make(List.empty[Next[M, E]])
+      } yield State(callsRef, nextRef)
+
+    val checkUnmetExpectations =
+      (state: State[M, E]) =>
+        state.callsRef.get
+          .filterOrElse[Any, Nothing, Any](_.isEmpty) { calls =>
+            val expectations = calls.map(call => call.method -> call.assertion)
+            ZIO.die(UnmetExpectationsException(expectations))
+          }
+
+    val makeEnvironment =
+      (state: State[M, E]) =>
+        for {
+          _    <- extract(state, self)
+          mock = Mock.make(state.callsRef)
+        } yield mockable.environment(mock)
+
+    for {
+      state <- Managed.make(makeState)(checkUnmetExpectations)
+      env   <- Managed.fromEffect(makeEnvironment(state))
+    } yield env
+  }
+
+  /**
+   * Returns empty expectation.
+   *
+   * This is required for the for-comprehension syntax.
+   */
+  @silent("parameter value f in method map is never used")
+  final def map[B](f: A => B): Expectation[M, E, Nothing] =
+    flatMap(_ => Empty)
+
+  /**
+   * A named alias for `*>`
+   */
+  final def zipRight[M1 <: M, E1 >: E, B](that: Expectation[M1, E1, B]): Expectation[M1, E1, B] = self *> that
+}
+
+object Expectation {
+
+  /**
+   * Returns a return expecation to fail with `E`.
+   */
+  final def failure[E](failure: E): Fail[Any, E] = Fail(_ => IO.fail(failure))
+
+  /**
+   * Maps the input arguments `I` to a return expecation to fail with `E`.
+   */
+  final def failureF[I, E](f: I => E): Fail[I, E] = Fail(i => IO.succeed(i).map(f).flip)
+
+  /**
+   * Effectfully maps the input arguments `I` to a return expecation to fail with `E`.
+   */
+  final def failureM[I, E](f: I => IO[E, Nothing]): Fail[I, E] = Fail(f)
+
+  /**
+   * Returns a return expecation to compute forever.
+   */
+  final def never: Succeed[Any, Nothing] = valueM(_ => IO.never)
+
+  /**
+   * Returns a return expecation to succeed with `Unit`.
+   */
+  final def unit: Succeed[Any, Unit] = value(())
+
+  /**
+   * Returns a return expecation to succeed with `A`.
+   */
+  final def value[A](value: A): Succeed[Any, A] = Succeed(_ => IO.succeed(value))
+
+  /**
+   * Maps the input arguments `I` to a return expecation to succeed with `A`.
+   */
+  final def valueF[I, A](f: I => A): Succeed[I, A] = Succeed(i => IO.succeed(i).map(f))
+
+  /**
+   * Effectfully maps the input arguments `I` to a return expecation to succeed with `A`.
+   */
+  final def valueM[I, A](f: I => IO[Nothing, A]): Succeed[I, A] = Succeed(f)
+
+  /**
+   * Implicitly converts Expectation to ZManaged mock environment.
+   */
+  implicit final def toManagedEnv[M, E, A](
```

Comment:
Should we put this in some separate object that can be optionally imported like `ZIOSyntax`? I understand why you want to have it for a more convenient API but using implicits, especially implicit conversions as opposed to implicit syntax and implicit conversions that are automatically imported when you import a top level object, seems like an anti-pattern for ZIO.

Alternatively. What if we just have syntax that adds a `provideExpectation` method on `ZIO` that would take an expectation, convert it to a mock environment, and then provide that to the effect. I think that would have the same benefits for usability but have less of the "magic" of implicit conversions.

Reply:
I've put it in `Expectation` on purpose. I'd like that implicit to be always available without the user haveing to import it. Basically, it allows to use the for-comprehension builder directly in `provideManaged`.

I don't think we should introduce a seperate `provideExpectation` for this.
The `Expectation` structure is supposed to be only a supporting builder, it's always going to be converted to managed environment. While I understand that haveing implicit magic is not always the best experience for the user, in this case I believe its beneficial:
* the user always wants to perform this conversion
* the user does not have to import anything, this conversion will always be in scope becouse it's placed in`Expectation`'s companion object
* the implicit conversion is total -> for every expectation it will yield a managed env
* there is no scenario where the user might want to perform this conversion on his own or replace it with a diffrent conversion

To summarize: this is to reduce the boilerplate and should be transparent to the user.

Reply:
Those are great points. I'll defer to you on this one. I was just thinking if you go back to the examples you currently have:

```scala
testM("expect call returning output") {
  val app     = clock.nanoTime
  val mockEnv = MockClock.nanoTime returns value(1000L)
  val result  = app.provideManaged(mockEnv)
  assertM(result, equalTo(1000L))
},
```

Where the implicit conversion is necessary to make the call to `provideManaged` work. With a `provideMock` syntax method you could instead write this as:

```scala
testM("expect call returning output") {
  val app     = clock.nanoTime
  val mockEnv = MockClock.nanoTime returns value(1000L)
  val result  = app.provideMock(mockEnv)
  assertM(result, equalTo(1000L))
}
```

At least to me that seems as convenient and maybe even has some value in clarifying intent that you are providing a mock implementation which behaves significantly differently than other implementations.

But happy to do what you think makes sense. I don't know this as well as you so maybe there are other use cases where this would not work as well.

Reply:
Let me sleep with this one, I'll think about possible use cases and if provideMock solves them.

Reply:
I think we should leave it as is. To introduce `provideMock` we'd have to modify `ZIO` which is in `core`, while `Expectation` is in `test`. We could ofcourse extend `ZIO` via implicit class, but... that would trade one implicit for another, so there is no point.

The other solution would be to move Expectation from `test` to `core`, but I feel thats not a good idea too.

Regarding the use cases I have some ideas what else could be improved, but nothing concrete yet, so let's merge it as is, and if I come up with some better design I'll put that into next PR.

Reply:
Okay. I think if we went with `provideMock` we would definitely want to do it as a syntax class the user could import, something like `zio.test.mock.syntax._` analogous to `zio.syntax._`. That would avoid changing anything in `core`.

You're right we would still have an implicit but I think an implicit syntax class is better than an implicit conversion from a ZIO perspective. It just gives the user an additional method they can call on something versus "magically" changing types. I think that is reflected in how there is some optional implicit syntax in ZIO today, for example in `ZIOSyntax` but to my knowledge there is no use of implicit conversions.

I'm okay with keeping it as is but just wanted to make sure I was clear about what I was proposing.